src/main/java/io/confluent/kafkarest/ConsumerReadTask.java

```diff
@@ -108,37 +110,50 @@ public boolean doPartialRead() {
       long startedIteration = parent.getConfig().getTime().milliseconds();
       final int requestTimeoutMs = parent.getConfig().getInt(
           KafkaRestConfig.CONSUMER_REQUEST_TIMEOUT_MS_CONFIG);
+      final long endTime = startedIteration + requestTimeoutMs;
+      final int itBackoff = parent.getConfig().getInt(
+        KafkaRestConfig.CONSUMER_ITERATOR_BACKOFF_MS_CONFIG);
       try {
-        // Read off as many messages as we can without triggering a timeout exception. The
-        // consumer timeout should be set very small, so the expectation is that even in the
-        // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
-        // check the elapsed time once this loop finishes.
-        while (iter.hasNext()) {
-          MessageAndMetadata<KafkaK, KafkaV> msg = iter.peek();
-          ConsumerRecordAndSize<ClientK, ClientV> recordAndSize = parent.createConsumerRecord(msg);
-          long roughMsgSize = recordAndSize.getSize();
-          if (bytesConsumed + roughMsgSize > maxResponseBytes) {
-            break;
+        while (parent.getConfig().getTime().milliseconds() < endTime) {
+
+          if (iter == null || !iter.hasNext()) {
+            // The consumer timeout should be set very small, so the expectation is that even in the
+            // worst case, num_messages * consumer_timeout << request_timeout, so it's safe to only
+            // check the elapsed time once this loop finishes.
+            ConsumerRecords<KafkaK, KafkaV> records = consumer.poll(itBackoff);
```

Comment:
@hachikuji my understanding is when records systematically appear in a topic and nobody retrieves them from REST proxy there is a risk of OutOfMemoryError with such approach. We can only bound records by the expiration timeout which gives us not much flexibility in configuration.
Could you please write a confirmation if I need to implement this approach.
Thank you!